tests/python/pants_test/util/test_objects.py

```diff
@@ -120,3 +168,139 @@ def test_unexpect_kwarg(self):
+
+    with self.assertRaises(TypeDecl.ConstructionError):
+      Union()
+
+    with self.assertRaises(TypeDecl.ConstructionError):
+      Union(str, 3)
+
+  def test_type_decl_composition(self):
+    str_type_decl = SimpleTypeDecl(str)
+    some_union_decl = Union(list, int)
+
+    composed_decl = str_type_decl.compose(some_union_decl)
+    self.assertTrue(composed_decl.matches_value(str('asdf')))
+    self.assertTrue(composed_decl.matches_value([]))
+    self.assertTrue(composed_decl.matches_value(3))
+    self.assertFalse(composed_decl.matches_value(type('asdf')))
+
+  def test_type_with_pred(self):
+    with self.assertRaises(TypeCheckError):
+      NonNegativeInt(value='asdf')
+
+    self.assertEqual(NonNegativeInt(value=3).value, 3)
+
+    with self.assertRaises(TypeCheckError):
+      NonNegativeInt(value=-1)
+
+
+class TypedDatatypeTest(BaseTest):
+
+  def test_class_construction(self):
+    # NB: typed_datatype subclasses declared at top level are the success cases
+    # here by not failing on import.
+
+    # If the type_name can't be converted into a suitable identifier, throw a
+    # ValueError.
+    with self.assertRaises(ValueError):
+      class NonStrType(typed_datatype(3, {'a': int})): pass
+
+    # This raises a TypeError because it doesn't provide a required argument.
+    with self.assertRaises(TypeError):
+      class NoFields(typed_datatype('NoFields')): pass
+
+    with self.assertRaises(TypedDatatypeClassConstructionError):
+      class NonDictFields(typed_datatype('NonDictFields', [
+          ('field', 'value'),
+      ])): pass
+
+    with self.assertRaises(TypedDatatypeClassConstructionError):
+      class NonTypeFields(typed_datatype('NonTypeFields', {'a': 3})): pass
+
+    with self.assertRaises(TypedDatatypeClassConstructionError):
+      class NonTypeUnionFields(typed_datatype('NonTypeUnionFields', {
+          'a': [str, 3],
+      })): pass
+
+    some_object = YetAnotherNamedTypedDatatype(
+      nothing_special=str('asdf'),
+      just_another_arg=3)
+    self.assertEqual(3, some_object.just_another_arg)
+
+    another_object = YetAnotherNamedTypedDatatype(
+      nothing_special=str('huh'),
+      just_another_arg=str('wow'))
+    self.assertEqual(str('wow'), another_object.just_another_arg)
+
+  def test_instance_construction(self):
+
+    some_val = SomeTypedDatatype(my_val=3)
+    self.assertIn('SomeTypedDatatype', repr(some_val))
+    self.assertIn('my_val', repr(some_val))
+    self.assertIn('3', repr(some_val))
+
+    union_val = UnionFieldTypedDatatype(an_arg=str('huh'))
+    self.assertIn('UnionFieldTypedDatatype', repr(union_val))
+    self.assertIn('an_arg', repr(union_val))
+    self.assertIn('huh', repr(union_val))
+
+    other_union_val = UnionFieldTypedDatatype(an_arg=3)
+    self.assertIn('UnionFieldTypedDatatype', repr(other_union_val))
+    self.assertIn('an_arg', repr(other_union_val))
+    self.assertIn('3', repr(other_union_val))
```

Comment:
I like that you've included tests of the repr. I'd like the assertions to be a bit more concrete though. I think testing the full repr would make it clearer what the repr is supposed to look like. How would you feel about that?

Particularly since the ideal for repr is to have a round-trippable string representation that if it were evaluated as python would recreate the object.

Reply:
I 100% agree. I have thoughts on stu's comment above (e.g. if there are no named args, then that would change the repr), but validating the repr in a structured way seems really appropriate for this use case.

Reply:
I mentioned this in a response to your earlier comment: check 133ba06 to see if the testing added there addresses your concern here.